### ISSUE TITLE
Skeletonize method="zhang" can produce skeleton mask with full submask 5x5.

### ISSUE THREAD

**Issue by vJlad** [scikit-image/scikit-image#8180]
### Description:

I've noticed that for my masks that were built from thresholding document pages, the default `skimage.morphology.skeletonize` produces invalid skeleton masks (based on the docs: "Skeletonization reduces binary objects to 1 pixel wide representations"). I cannot share my documents, so I tried to reproduce a toy example with a large full submask.

<img width="1209" height="407" alt="Image" src="https://github.com/user-attachments/assets/49d5b8f8-3488-4ded-85a8-12779cfc45ba" />

### Way to reproduce:

```python

import skimage.morphology
from matplotlib import pyplot as plt
import cv2
import numpy as np
from tqdm.auto import tqdm

np.random.seed(2312)
for i in tqdm(range(10000000)):
    mask = np.random.random((30, 30)) > np.random.random()
    mask = cv2.morphologyEx(
        mask.astype(np.uint8) * 255,
        cv2.MORPH_OPEN,
        kernel=cv2.getStructuringElement(cv2.MORPH_RECT, (3, 3)),
    )
    s1 = skimage.morphology.skeletonize(mask, method="zhang")
    if np.any(
        cv2.morphologyEx(s1.astype(np.uint8) * 255, cv2.MORPH_HITMISS, np.ones((5, 5)))
    ):
        s2 = skimage.morphology.skeletonize(mask, method="lee")
        plt.figure(figsize=(15, 5))
        plt.subplot(1, 3, 1)
        plt.title("zhang")
        plt.imshow(s1)
        plt.subplot(1, 3, 2)
        plt.title("lee")
        plt.imshow(s2)
        plt.subplot(1, 3, 3)
        plt.title("initial mask")
        plt.imshow(mask)
        plt.show()
        break

```

### Version information:

```Shell
3.13.13 (main, May  4 2026, 21:09:13) [Clang 22.1.3 ]
Linux-6.6.87.2-microsoft-standard-WSL2-x86_64-with-glibc2.39
scikit-image version: 0.26.0
numpy version: 2.4.4
```

**Comment by stefanv** [comment]
Thanks for the report @vJlad.

Here's the skimage-only version of the reproducer:

```py
import numpy as np
import skimage as ski
from matplotlib import pyplot as plt

np.random.seed(2312)
for i in range(10_000_000):
    mask = np.random.random((30, 30)) > np.random.random()
    mask = ski.morphology.binary_opening(mask, footprint=np.ones((3, 3)))
    s1 = ski.morphology.skeletonize(mask, method="zhang")
    if np.any(ski.morphology.erosion(s1, footprint=np.ones((5, 5)))):
        s2 = ski.morphology.skeletonize(mask, method="lee")
        fig, axes = plt.subplots(1, 3, figsize=(15, 5))
        for ax, img, title in zip(
            axes,
            [s1, s2, mask],
            ["zhang", "lee", "initial mask"],
        ):
            ax.imshow(img)
            ax.set_title(title)
        plt.savefig('skeletonize-bug.png')
        break
```

**Comment by stefanv** [comment]
This exploration sent me down quite the rabbit hole.

Turns out, the behavior you are seeing in Zhang-Suen is known. In fact, scikit-image does not implement the original paper exactly; we've already made some modifications to improve its results. But from what I've read, post-processing is often recommended.

There are, however, newer algorithms that produce much better results, of which I'll attach two:

(Input, 1994 Bertrand et al., neural net: Menten et al.)

<img width="170" alt="Image" src="https://github.com/user-attachments/assets/2ad09169-c097-4f48-824b-9bacc3edbee9" />

<img width="100" alt="Image" src="https://github.com/user-attachments/assets/ea5a0c84-338d-4266-a96e-e11edf5c337e" />

<img width="170" alt="Image" src="https://github.com/user-attachments/assets/18fcefc6-eaec-4f28-97f8-ca32a9ddeaa4" />

Software that implements these:

- [GDtal](https://www.dgtal.org/) — LGPL
- [pink](https://github.com/hugues-talbot/pin) — CeCILL — @hugues-talbot and [Michel Couprie](https://perso.esiee.fr/~coupriem/index-eng.html)
- [Skeletonization for Gradient-based Optimization](https://github.com/martinmenten/skeletonization-for-gradient-based-optimization) — no license — @martinmenten

I suspect that we could get Martin's code to run without GPU, but would have to check.

If anyone can suggest a leading skeletonization algorithm to include in scikit-image, that would be very helpful! Also, adding 3D is quite appealing (see references below).

**References**

- [Menten et al., A skeletonization algorithm for gradient-based optimization](https://arxiv.org/pdf/2309.02527)
- [Bertrand & Couprie, Parallel skeletonization algorithms in the cubic grid based on
critical kernels](https://enpc.hal.science/hal-01613008v1/file/CriticalSynthese.pdf)
- [Bertrand & Couprie, Powerful Parallel and Symmetric 3D Thinning Schemes Based on Critical Kernels](https://hal.science/file/index/docid/828450/filename/3Dschemes-hal.pdf)
- [Bertrand, Simple points, topological numbers and geodesic neighborhoods in cubic grids](https://www.sciencedirect.com/science/article/pii/0167865594900329) — paywalled, sorry.
- [Zhang & Suen, A Fast Parallel Algorithm for Thinning Digital Patterns](https://dl.acm.org/doi/epdf/10.1145/357994.358023) — the OG

The [ImageJ documentation page on skeletonization](https://imagej.net/plugins/analyze-skeleton/) is also instructive—they've clearly put some work into it [[3D version](https://imagej.net/plugins/skeletonize3d)].

**Comment by martinmenten** [comment]
Thanks for noticing our work on skeletonization for gradient-based optimization and including me in this discussion!

I have gladly added an MIT license to our code - I should have done this a long time ago. However, I am not sure whether my implementation of Bertrand’s method is useful for a library like scikit-image for the following two reasons:

**Reason 1:** The aim of our work was to research whether skeletonization algorithms can be integrated with backpropagation and deep learning. As a result, I had to replace certain algorithmic elements that do not work well with backpropagation. For example, I avoided large look-up tables similar to those used in the method by Lee et al., or I replaced slicing and summing with equivalent kernel-based operations.

To illustrate this, consider a function from my code that is used to determine whether a point has a single or no neighbor in its 26-neighborhood at all:

```
def _single_neighbor_check(self, img):
    """
    Function that characterizes points as endpoints if they have a single neighbor or no neighbor at all.
    """

    # Check that number of ones in twentysix-neighborhood is exactly 0 or 1
    K = torch.tensor([[[1.0, 1.0, 1.0],
                       [1.0, 1.0, 1.0],
                       [1.0, 1.0, 1.0]],
                      [[1.0, 1.0, 1.0],
                       [1.0, 0.0, 1.0],
                       [1.0, 1.0, 1.0]],
                      [[1.0, 1.0, 1.0],
                       [1.0, 1.0, 1.0],
                       [1.0, 1.0, 1.0]]], device=img.device).view(1, 1, 3, 3, 3)

    num_twentysix_neighbors = F.conv3d(img, K)
    condition1 = F.hardtanh(-(num_twentysix_neighbors - 2), min_val=0, max_val=1) # 1 or fewer neighbors
    
    return condition1
```

This could be much more elegantly and probably also more efficiently done by just using slicing and a single sum operation.

My skeletonization function also runs for a fixed number of iterations that is manually set. This is because I wanted to avoid checking whether a processed image is still altered as a result of the skeletonization operation. Furthermore, I currently handle 2D images by simply padding them with zero slices and subsequently processing the pseudo-3D input.

As a result of the aforementioned issues, my code is rather inefficient, especially with regard to memory usage. While all of the aforementioned issues can probably be fixed with enough effort put into reworking the code, there may be other implementations of the skeletonization algorithm by Bertrand et al. that are written with a much higher focus on computational efficiency.

**Reason 2**: Moreover, there could be skeletonization algorithms other than the Bertrand’s method that are more suitable for the design goals of scikit-image. The reason I chose Bertrand’s method for my work, is that it is fully topologically correct and expressible with deep-learning-compatible operations (see point above).

I can very much sympathize with the figurative “rabbit hole” mentioned by @stefanv that is skeletonization algorithms for digital images. Not only have there been many published algorithms by now, but the topological skeleton of a digital image is a not clearly defined concept. Personally, I believe the version of Lee et al. is very close to an optimal solution for scikit-image, although there used to be a bug in it when processing 3D cases (see also #3757 ).

**Comment by plelievre** [comment]
Hi, I have written a differentiable version of _Zhang-Suen, 1984_ in PyTorch (just as @martinmenten did), but the results are different from the version implemented in scikit-image. I think that @stefanv gave an explanation:

> Turns out, the behavior you are seeing in Zhang-Suen is known. In fact, scikit-image does not implement the original paper exactly; we've already made some modifications to improve its results. But from what I've read, post-processing is often recommended.

However, as some papers refer to _Zhang-Suen, 1984_ when they are actually using scikit-image's version, it would be nice to clarify in the doc the operated modifications in terms of rules.

Thank you.
